app/src/main/java/org/thoughtcrime/securesms/util/MediaUtil.java

```diff
@@ -0,0 +1,477 @@
+      return false;
+    }
+
+    if ("com.android.providers.media.documents".equals(uri.getAuthority())) {
+      return uri.getLastPathSegment().contains("video");
+    } else if (uri.toString().startsWith(MediaStore.Video.Media.EXTERNAL_CONTENT_URI.toString())) {
+      return true;
+    } else if (uri.toString().startsWith("file://") &&
+               MediaUtil.isVideo(URLConnection.guessContentTypeFromName(uri.toString()))) {
+      return true;
+    } else if (PartAuthority.isAttachmentUri(uri) && MediaUtil.isVideoType(PartAuthority.getAttachmentContentType(context, uri))) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  @WorkerThread
+  public static @Nullable Bitmap getVideoThumbnail(@NonNull Context context, @Nullable Uri uri, long timeUs) {
+    if (uri == null) {
+      return null;
+    } else if ("com.android.providers.media.documents".equals(uri.getAuthority())) {
+      long videoId = Long.parseLong(uri.getLastPathSegment().split(":")[1]);
+
+      return MediaStore.Video.Thumbnails.getThumbnail(context.getContentResolver(),
+                                                      videoId,
+                                                      MediaStore.Images.Thumbnails.MINI_KIND,
+                                                      null);
+    } else if (uri.toString().startsWith(MediaStore.Video.Media.EXTERNAL_CONTENT_URI.toString())) {
+      long videoId = Long.parseLong(uri.getLastPathSegment());
+
+      return MediaStore.Video.Thumbnails.getThumbnail(context.getContentResolver(),
+                                                      videoId,
+                                                      MediaStore.Images.Thumbnails.MINI_KIND,
+                                                      null);
+    } else if (uri.toString().startsWith("file://") &&
+               MediaUtil.isVideo(URLConnection.guessContentTypeFromName(uri.toString()))) {
+      return ThumbnailUtils.createVideoThumbnail(uri.toString().replace("file://", ""),
+                                                 MediaStore.Video.Thumbnails.MINI_KIND);
+    } else if (Build.VERSION.SDK_INT >= 23   &&
+               BlobProvider.isAuthority(uri) &&
+               MediaUtil.isVideo(BlobProvider.getMimeType(uri)))
+    {
+      try {
+        MediaDataSource source = BlobProvider.getInstance().getMediaDataSource(context, uri);
+        return extractFrame(source, timeUs);
+      } catch (IOException e) {
+        Log.w(TAG, "Failed to extract frame for URI: " + uri, e);
+      }
+    } else if (Build.VERSION.SDK_INT >= 23        &&
+               PartAuthority.isAttachmentUri(uri) &&
+               MediaUtil.isVideoType(PartAuthority.getAttachmentContentType(context, uri)))
+    {
+      try {
+        AttachmentId    attachmentId = PartAuthority.requireAttachmentId(uri);
+        MediaDataSource source       = DatabaseFactory.getAttachmentDatabase(context).mediaDataSourceFor(attachmentId);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *source*. Currently, there are execution paths that do not contain closure statements, for example, when MediaMetadataRetrieverUtil.setDataSource (in method extractFrame) throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).